File: backend/src/lib/searchText.ts
import TinySegmenter from "tiny-segmenter";

const segmenter = new TinySegmenter();

const TOKEN_PATTERN =
  /[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}ー]+|[\p{L}\p{N}]+(?:[_-][\p{L}\p{N}]+)*/gu;
const JAPANESE_TOKEN_PATTERN = /[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}]/u;
const SINGLE_KANA_PATTERN = /^[ぁ-ゖァ-ヺー]$/u;

const ENGLISH_STOPWORDS = new Set([
  "a",
  "an",
  "and",
  "are",
  "for",
  "from",
  "how",
  "into",
  "our",
  "that",
  "the",
  "this",
  "use",
  "with",
  "your"
]);

const JAPANESE_STOPWORDS = new Set([
  "これ",
  "それ",
  "あれ",
  "こと",
  "もの",
  "ため",
  "よう",
  "です",
  "ます",
  "した",
  "して",
  "する",
  "ある",
  "いる",
  "この",
  "その",
  "ですか",
  "ますか",
  "ください",
  "お願いします"
]);

function extractTokenCandidates(input: string) {
  const normalized = input.normalize("NFKC").replace(/\s+/g, " ").trim();
  if (!normalized) {
    return [];
  }

  const tokens: string[] = [];

  for (const segment of segmenter.segment(normalized)) {
    const matches = segment.match(TOKEN_PATTERN);
    if (matches?.length) {
      tokens.push(...matches);
    }
  }

  return tokens;
}

function cleanToken(token: string) {
  const normalized = token.normalize("NFKC").trim().toLowerCase();
  if (!normalized) {
    return null;
  }

  if (JAPANESE_TOKEN_PATTERN.test(normalized)) {
    if (SINGLE_KANA_PATTERN.test(normalized) || JAPANESE_STOPWORDS.has(normalized)) {
      return null;
    }
    return normalized;
  }

  if (normalized.length < 2 || ENGLISH_STOPWORDS.has(normalized)) {
    return null;
  }

  return normalized;
}

export function buildSearchText(...parts: Array<string | null | undefined>) {
  const tokens: string[] = [];

  for (const part of parts) {
    if (!part) {
      continue;
    }

    for (const candidate of extractTokenCandidates(part)) {
      const token = cleanToken(candidate);
      if (token) {
        tokens.push(token);
      }
    }
  }

  return tokens.join(" ");
}

export function tokenizeSearchTerms(input: string, limit = 12) {
  const terms: string[] = [];
  const seen = new Set<string>();

  for (const candidate of extractTokenCandidates(input)) {
    const token = cleanToken(candidate);
    if (!token || seen.has(token)) {
      continue;
    }

    seen.add(token);
    terms.push(token);

    if (terms.length >= limit) {
      break;
    }
  }

  return terms;
}

function escapeFtsToken(token: string) {
  return `"${token.replace(/"/g, "\"\"")}"`;
}

export function toFtsQuery(input: string) {
  const tokens = tokenizeSearchTerms(input);
  if (!tokens.length) {
    return "";
  }

  return tokens.map((token) => `${escapeFtsToken(token)}*`).join(" OR ");
}

File: internal/search/searchtext.go
package search

import (
	"regexp"
	"strings"
	"unicode/utf8"

	"golang.org/x/text/unicode/norm"
)

// Port of backend/src/lib/searchText.ts. The application pre-tokenizes text in
// this layer before inserting into / querying the FTS5 tables, so the Go output
// must match the JS output exactly (verified by searchtext_test.go).

// JS: /[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}ー]+|[\p{L}\p{N}]+(?:[_-][\p{L}\p{N}]+)*/gu
var tokenPattern = regexp.MustCompile(`[\p{Han}\p{Hiragana}\p{Katakana}ー]+|[\p{L}\p{N}]+(?:[_-][\p{L}\p{N}]+)*`)

// JS: /[\p{Script=Han}\p{Script=Hiragana}\p{Script=Katakana}]/u
var japaneseTokenPattern = regexp.MustCompile(`[\p{Han}\p{Hiragana}\p{Katakana}]`)

// JS: /^[ぁ-ゖァ-ヺー]$/u
var singleKanaPattern = regexp.MustCompile(`^[ぁ-ゖァ-ヺー]$`)

// JS: .replace(/\s+/g, " ")
var whitespacePattern = regexp.MustCompile(`\s+`)

var englishStopwords = map[string]bool{
	"a": true, "an": true, "and": true, "are": true, "for": true,
	"from": true, "how": true, "into": true, "our": true, "that": true,
	"the": true, "this": true, "use": true, "with": true, "your": true,
}

var japaneseStopwords = map[string]bool{
	"これ": true, "それ": true, "あれ": true, "こと": true, "もの": true,
	"ため": true, "よう": true, "です": true, "ます": true, "した": true,
	"して": true, "する": true, "ある": true, "いる": true, "この": true,
	"その": true, "ですか": true, "ますか": true, "ください": true, "お願いします": true,
}

func extractTokenCandidates(input string) []string {
	normalized := norm.NFKC.String(input)
	normalized = whitespacePattern.ReplaceAllString(normalized, " ")
	normalized = strings.TrimSpace(normalized)
	if normalized == "" {
		return nil
	}

	var tokens []string
	for _, segment := range Segment(normalized) {
		matches := tokenPattern.FindAllString(segment, -1)
		tokens = append(tokens, matches...)
	}
	return tokens
}

// cleanToken returns the cleaned token and whether it should be kept.
func cleanToken(token string) (string, bool) {
	normalized := strings.ToLower(strings.TrimSpace(norm.NFKC.String(token)))
	if normalized == "" {
		return "", false
	}

	if japaneseTokenPattern.MatchString(normalized) {
		if singleKanaPattern.MatchString(normalized) || japaneseStopwords[normalized] {
			return "", false
		}
		return normalized, true
	}

	if utf8.RuneCountInString(normalized) < 2 || englishStopwords[normalized] {
		return "", false
	}
	return normalized, true
}

// BuildSearchText tokenizes the given parts into a space-joined string that is
// stored in the FTS5 tables. Mirrors buildSearchText(...parts).
func BuildSearchText(parts ...string) string {
	var tokens []string
	for _, part := range parts {
		if part == "" {
			continue
		}
		for _, candidate := range extractTokenCandidates(part) {
			if token, ok := cleanToken(candidate); ok {
				tokens = append(tokens, token)
			}
		}
	}
	return strings.Join(tokens, " ")
}

// TokenizeSearchTerms returns up to limit unique tokens from input.
// Mirrors tokenizeSearchTerms(input, limit=12).
func TokenizeSearchTerms(input string, limit int) []string {
	if limit <= 0 {
		limit = 12
	}
	terms := make([]string, 0, limit)
	seen := make(map[string]bool)

	for _, candidate := range extractTokenCandidates(input) {
		token, ok := cleanToken(candidate)
		if !ok || seen[token] {
			continue
		}
		seen[token] = true
		terms = append(terms, token)
		if len(terms) >= limit {
			break
		}
	}
	return terms
}

func escapeFtsToken(token string) string {
	return `"` + strings.ReplaceAll(token, `"`, `""`) + `"`
}

// ToFtsQuery converts input into an FTS5 MATCH query. Mirrors toFtsQuery(input).
func ToFtsQuery(input string) string {
	tokens := TokenizeSearchTerms(input, 12)
	if len(tokens) == 0 {
		return ""
	}
	parts := make([]string, 0, len(tokens))
	for _, token := range tokens {
		parts = append(parts, escapeFtsToken(token)+"*")
	}
	return strings.Join(parts, " OR ")
}
